src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -45,11 +47,14 @@ private void AddExceptionDiagnostic(Diagnostic diagnostic)
         /// </summary>
         public async Task<ImmutableArray<Diagnostic>> GetAnalyzerDiagnosticsAsync()
         {
+            _driver.StartCompleteAnalysis(_analyzedSyntaxTrees, _cancellationToken);
+
             // Invoke GetDiagnostics to populate the compilation's CompilationEvent queue.
             // Discard the returned diagnostics.
             _compilation.GetDiagnostics(_cancellationToken);
```

Comment:
This essentially means that no analyzer diagnostics will be returned till entire compilation events are generated, even though some of those events might have been processed and corresponding analyzer diagnostics reported. Should GetAnalyzerDiagnosticsAsync take a flag "forceCompleteCompilation" which guards this call to Compilation.GetDiagnostics?